src/test/netmsg_tests.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) 2022-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <net_processing.h>
+#include <netaddress.h>
+#include <protocol.h>
+#include <test/util/logging.h>
+#include <test/util/net.h>
+#include <test/util/setup_common.h>
+#include <util/time.h>
+#include <version.h>
+
+#include <boost/test/unit_test.hpp>
+
+#include <chrono>
+#include <string>
+#include <unordered_map>
+
+BOOST_FIXTURE_TEST_SUITE(netmsg_tests, NetTestingSetup)
+
+BOOST_AUTO_TEST_CASE(initial_messages_exchange)
+{
+    std::unordered_map<std::string, size_t> count_sent_messages;
+    auto pipes = m_sockets_pipes.PopFront();
+
+    // Wait for all messages due to the initial handshake to be Send() to the socket.
+    // The FEEFILTER is the last one, so quit when we get that.
+    for (;;) {
+        auto msg = pipes->send.GetNetMsg();
+        if (msg.has_value()) {
+            count_sent_messages[msg->m_type]++;
+            if (msg->m_type == NetMsgType::FEEFILTER) {
+                break;
+            }
+        } else {
+            break;
+        }
+    }
+
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::VERSION], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::WTXIDRELAY], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::SENDADDRV2], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::VERACK], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::GETADDR], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::SENDCMPCT], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::PING], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::GETHEADERS], 1);
+    BOOST_CHECK_EQUAL(count_sent_messages[NetMsgType::FEEFILTER], 1);
+}
+
+BOOST_AUTO_TEST_CASE(addr)
+{
+    auto pipes = m_sockets_pipes.PopFront();
+    std::vector<CAddress> addresses{5};
+
+    DebugLogHelper debug_log{strprintf("Received addr: %u addresses", addresses.size())};
+    pipes->recv.PushNetMsg(NetMsgType::ADDRV2, addresses);
+    debug_log.EndAndThrowIfNotFound(30s);
+}
+
+BOOST_AUTO_TEST_CASE(getblocks)
+{
+    auto pipes = m_sockets_pipes.PopFront();
+    std::vector<uint256> hashes{5};
+    CBlockLocator block_locator{std::move(hashes)};
+    uint256 hash_stop;
+
+    DebugLogHelper debug_log{"getblocks -1 to end"};
+    pipes->recv.PushNetMsg(NetMsgType::GETBLOCKS, block_locator, hash_stop);
+    debug_log.EndAndThrowIfNotFound(30s);
+}
+
+BOOST_AUTO_TEST_CASE(ping)
```

Comment:
This is inspired by https://github.com/bitcoin/bitcoin/pull/25515

Reply:
Yes, both testing in isolation and e2e is good to have. The difference between the tests included in this PR and the functional tests is that when the test code runs inside the same process:

* it has greater control and observability - the test can read any global or member variable and can call arbitrary functions 
* it can do fuzzing

I think all those tests are complementary, not mutually exclusive.